packages/react-native/ReactCommon/react/nativemodule/idlecallbacks/NativeIdleCallbacks.cpp

```diff
@@ -0,0 +1,126 @@
+      size_t /* unused */) noexcept -> jsi::Value {
+        if (runtimeScheduler->getShouldYield()) {
+          return jsi::Value(runtime, 0.0);
+        }
+              
+        auto now = runtimeScheduler->now();
+        double remainingTime = std::chrono::duration_cast<std::chrono::milliseconds>(deadline - now).count();
+        return jsi::Value(runtime, std::max(remainingTime, 0.0));
+    });
+}
+
+NativeIdleCallbacks::NativeIdleCallbacks(std::shared_ptr<CallInvoker> jsInvoker)
+  : NativeIdleCallbacksCxxSpec(std::move(jsInvoker)) {}
+
+CallbackHandle NativeIdleCallbacks::requestIdleCallback(
+  jsi::Runtime& runtime,
+  jsi::Function callback,
+  std::optional<NativeRequestIdleCallbackOptions> options) {
+  using namespace std::chrono_literals;
+
+  auto binding = RuntimeSchedulerBinding::getBinding(runtime);
+  auto runtimeScheduler = binding->getRuntimeScheduler();
+
+  // handle timeout parameter
+  std::optional<double> timeout;
+  if (options.has_value()) {
+    timeout = options.value().timeout;
+  }
+
+  auto userCallback = std::make_shared<jsi::Function>(std::move(callback));
+
+  auto wrappedCallback = jsi::Function::createFromHostFunction(
+    runtime,
+    jsi::PropNameID::forAscii(runtime, "requestIdleCallback$bodyWrapper"),
+    1,
+    [runtimeScheduler, cb = std::move(userCallback)](
+      jsi::Runtime& runtime,
+      const jsi::Value& /* unused */,
+      const jsi::Value* args,
+      size_t /* unused */
+    ) noexcept -> jsi::Value {
+      auto didTimeout = args[0].getBool();
+      // Below is a partial solution and does not comply to WHATWG event-loop standards
+      auto deadline = runtimeScheduler->now() + 50ms;
+
+      jsi::Object idleDeadline {runtime};
+      idleDeadline.setProperty(runtime, "didTimeout", didTimeout);
+      idleDeadline.setProperty(
+        runtime,
+        "timeRemaining",
+        makeTimeRemainingFunction(runtime, runtimeScheduler, deadline)
+      );
+
+      return cb->call(runtime, std::move(idleDeadline));
+    }
+  );
+
+  auto timeoutDuration = std::chrono::duration<double, std::milli>(timeout.value_or(0.0));
+  auto task = runtimeScheduler->scheduleIdleTask(std::move(wrappedCallback),
+                                                 std::chrono::duration_cast<RuntimeSchedulerTimeout>(timeoutDuration));
+
+  if (task == nullptr) {
+    throw new jsi::JSError(runtime, "requestIdleCallback is not supported in legacy runtime scheduler");
+  }
+
+  jsi::Object taskHandle {runtime};
+  auto taskNativeState = std::make_shared<IdleTaskRef>(task);
+  taskHandle.setNativeState(runtime, std::move(taskNativeState));
+
+  return taskHandle;
+}
+
+void NativeIdleCallbacks::cancelIdleCallback(jsi::Runtime& runtime, jsi::Object handle) {
+  auto binding = RuntimeSchedulerBinding::getBinding(runtime);
+  auto runtimeScheduler = binding->getRuntimeScheduler();
+
+  if (!handle.hasNativeState(runtime)) {
+    throw new jsi::JSError(runtime, "Invalid idle task handle");
```

Comment:
I just checked and `cancelIdleCallback` never throws on Web, so we should probably do the same here.